<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="GAG - Genome Annotation Generator : Console-based application to read, sanitize, annotate and modify genomic data. Can export an NCBI .tbl file of annotations on a genome." />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>GAG - Genome Annotation Generator</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/BrianReallyMany/GAG">View on GitHub</a>

          <h1 id="project_title">GAG - Genome Annotation Generator</h1>
          <h2 id="project_tagline">Console-based application to read, sanitize, annotate and modify genomic data. Can export an NCBI .tbl file of annotations on a genome.</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/BrianReallyMany/GAG/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/BrianReallyMany/GAG/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a name="introduction" class="anchor" href="#introduction"><span class="octicon octicon-link"></span></a>Introduction</h1>

<h4>License</h4>
<p>GAG is released under the <a href="http://opensource.org/licenses/MIT">MIT License</a>.</p>

<h4>Purpose</h4>
<p>GAG can read a genome and write it to the NCBI's .tbl format. This is the main purpose of the program. However, if you're writing a .tbl file, there's a good chance that your genome is destined to encounter <a href="https://www.ncbi.nlm.nih.gov/genbank/tbl2asn2">tbl2asn</a> further downstream. In anticipation of this meeting, GAG includes a (growing) number of options to remove questionable features or trim erroneous regions from your genome. These options are outlined in the <a href="#advanced-usage">Advanced Usage</a> secion.</p>
<h4>Genome</h4>
<p>What is a genome? Great question. A genome can be many things to many people. To GAG, a genome is a .fasta file and a .gff file. More specifically, a genome is a folder containing two files called 'genome.fasta' and 'genome.gff'. Anything else in the folder will be ignored.</p>
<h4>Fasta File</h4>
<p>You must provide GAG with a .fasta file of nucleic acid sequence data. (More info <a href="https://en.wikipedia.org/wiki/FASTA_format">here</a>.) It is important to note that GAG is only expecting to find 'A', 'C', 'G', 'T' or 'N' in your sequences. Mixed bases ('R', 'M', etc.) or invalid bases ('Q', 'ã…Š', etc.) are converted to 'N' for the purpose of protein translation, start/stop codon detection or whatever else we might dream up.
<h4>GFF File</h4>
<p>For best results, use a valid .gff file. The .gff specification can be found at the <a href="http://www.sequenceontology.org/gff3.shtml">Sequence Ontology</a> homepage, and there is a validator <a href="http://modencode.oicr.on.ca/cgi-bin/validate_gff3_online">here</a>. (For unpredictable results, go ahead and try to load an invalid file. It's been known to work. Who knows?)</p>
<p>GAG can read and write functional annotations. All of this happens in the "attributes" column of the file. The "ID" field is required to be unique for each feature; the "Parent" field is required for all non-gene features; everything else is gravy. For example, add a unique "Name" field to your gene and it will be included in your .tbl file. Attach "Dbxref" or "Ontology_term" fields to your mRNA feature and these annotations will be added when you write a .tbl file.</p>
<h4>Installation</h4>
<p>GAG is written in Python. Assuming you have Python (2, not 3) installed, you can download the zipped source code from the link at the top of this page, extract it and run it. The command to start the program is <pre><code>python gag.py</code></pre></p>
<p>It's recommended (though not required) to run GAG from inside the folder that contains your genome. To make this easier, you might consider adding GAG to your $PATH variable. One way to do that is to type the following from a Linux prompt:
<pre><code>export PATH=$PATH:/path/to/wherever/you/installed/GAG</code></pre></p>
<p>Adding the above line to your '~/.bash_profile' file will ensure that GAG is always at your fingertips.</p>
<h4>Help</h4>
<p>GAG features contextual help spread out amongst a number of task-oriented <a href="#nested-subconsoles">subconsoles</a>. At any given point, type <code>help</code> to see a list of available commands, and type <code>help &lt;name_of_command&gt;</code> to get instructions about a specific command.</p>
</p>Please note that the available commands will differ depending on which subconsole you are in. Frustrated? Lost? Type <code>home</code> to get back to the main console.</p>
<p>If that doesn't cut it, and this document doesn't answer your questions, you can visit our <a href="https://groups.google.com/forum/#!forum/gag_support">Support Page</a>.
<h4><a name="citing-gag" class="anchor" href="#citing-gag"></a>Citing GAG</h4>
<p>Please cite as follows:</p>
<blockquote>Geib, S., DeRego, T., &amp; Hall, B. (2014). GAG: the Genome Annotation Generator (Version 1.0) [Software]. Available from <a href="http://brianreallymany.github.io/GAG/">http://brianreallymany.github.io/GAG</a></blockquote>
<h1><a name="basic-usage" class="anchor" href="#basic-usage"><span class="octicon octicon-link"></span></a>Basic Usage</h1>



<p>In this exciting vignette, you will create a .tbl file (along with cds, mrna and protein .fasta files) from the .fasta and .gff files provided in the 'walkthrough/basic' directory.</p>
<p>GAG writes a number of (lightweight) history, config, stats and log files while you run it. For best results, run GAG from within the directory that contains your genome. This will keep all those files in one place.</p>
<p>The following walkthroughs all assume that you have taken this advice; if you run GAG from another folder, you should adjust the path in the 'load' command.</p>
<p>When GAG starts up, you'll see an exciting load screen and the GAG prompt. It looks like this:
<pre><code>GAG&gt; </pre></code>
<p>(In the examples that follow, the prompt is just there to make the tutorial resemble your screen. You don't need to type it. The only time you should ever need to type 'GAG' in all caps is when you <a href="#citing-gag">cite it</a> in your amazing paper.)</p>

<h5>
<a name="read-in-genome" class="anchor" href="#read-in-genome"><span class="octicon octicon-link"></span></a>Read in genome</h5>

<pre><code>GAG&gt; load .
</code></pre>

<h5>
<a name="write-output-files" class="anchor" href="#write-output-files"><span class="octicon octicon-link"></span></a>Write output files</h5>

<pre><code>GAG&gt; write newfolder
</code></pre>

<h5>
<a name="go-on-your-merry-way" class="anchor" href="#go-on-your-merry-way"><span class="octicon octicon-link"></span></a>Go on your merry way</h5>

<pre><code>GAG&gt; exit
</code></pre>

<p>You now have a folder in your working directory called 'newfolder'. It contains a .gff, several .fastas and your shiny new .tbl file. Congratulations!</p>
<h1><a name="advanced-usage" class="anchor" href="#advanced-usage"><span class="octicon octicon-link"></span></a>Advanced Usage</h1>
<p>The following two vignettes demonstrate some of GAG's capabilities for removing questionable features of your genome, fixing mistakes, and flagging features for further investigation.</p>

<h3>Removing Terminal 'N's</h3>
<p>The NCBI will not accept submissions containing sequences that begin or end with unknown bases. For example, take a look at the file 'walkthrough/terminal_ns/genome.fasta'. It is crawling with unknown bases. This type of sequence will earn your genome submission a big read REJECTED stamp.</p>
<p>GAG can check each sequence in your .fasta for 'N's at the beginning or the end. If it finds any, it will remove them from the sequence and (if the bases were at the beginning of the sequence) update the indices of all features on the sequence accordingly. If by chance you have a gene that starts on one of these unknown bases, it will be removed. Such is life.</p>
<p>To harness this awesome power, first start GAG and load the genome contained in the 'walkthrough/terminal_ns' folder. Then type <code>fix</code>. This takes you to the 'GAG FIX' subconsole. Your first subconsole! Nice work. Type <code>help</code> to see commands available within this subconsole. You should be looking at something like this:
<pre><code>GAG&gt; fix
----------------------------------------

This is the GAG FIX menu.
You can apply the following fixes: terminal_ns, start_stop_codons.
(You can type 'home' at any time to return to the main GAG console.)
Type the name of a fix to enable it. Type 'all' to enable everything.

terminal_ns, start_stop_codons or all?

GAG FIX&gt; help

Documented commands (type help <topic>):
========================================
all  home  start_stop_codons  terminal_ns

Undocumented commands:
======================
help

GAG FIX&gt;
</code></pre></p>
<p>Type <code>terminal_ns</code> (now is a great time to try out GAG's <a href="#tab-completion">tab completion</a> feature). Press enter. If all went well, you should see this:<pre><code>GAG FIX&gt; terminal_ns

Terminal Ns will now be fixed.

----------------------------------------
GAG&gt;</code></pre>
<p>Note that no changes have been made to the genome at this point. Only when you view a feature, ask for info or write the genome to a new folder will this fix be applied.</p>
<p>To confirm that something actually happened, write the modified genome to a folder:
<pre><code>GAG&gt; write terminal_ns_removed</code></pre></p>
<p>Exit and have a look at the results. If you open the file 'terminal_ns_removed/genome.fasta' you can see that it contains no 'N's. Additionally, the indices of each feature in 'terminal_ns_removed/genome.gff' are updated to correspond to the new, cleaned up sequence.</p>



<h1><a name="additional-notes" class="anchor" href="#additional-notes"><span class="octicon octicon-link"></span></a>Additional Notes</h1>
<p>Users who anticipate a long, fruitful career as GAG wizards should familiarize themselves with these GAG Facts and Features.</p>

<h3><a name="tab-completion" class="anchor" href="#tab-completion"></a>Tab Completion</h3>
<p>GAG implements contextual tab completion. This is a fancy way of saying that sometimes hitting the Tab key will finish your word and sometimes it won't. To be more specific, tab completion works on available commands (depending on your subconsole), but not arguments. Full-on psychic power tab completion is on a <a href="https://github.com/BrianReallyMany/GAG/issues/63">TODO list</a>.</p>

<h3><a name="nested-subconsoles" class="anchor" href="#nested-subconsoles"></a>Nested Subconsoles</h3>
<p>The prompt you see when you first start the program simply says <code>GAG> </code>. This is the main GAG console. If you type <code>help</code> you will see a list of commands. We have tried to keep this list, and the command names themselves, as short as possible. However, this is a program that lets you do things like "remove all mRNAs containing an intron shorter than 10 base pairs in length, while flagging any gene longer than 10,000 base pairs." How to offer that functionality without having umpteen commands all written in LengthyAndTypographicalErrorProneCamelCase? That's right, all together now: nested subconsoles.</p>
<p>To take the first of the commands above as an example, the first thing you must do is announce your intention to <code>remove</code> something. This takes you to the <code>GAG REMOVE&gt; </code> subconsole. Typing <code>help</code> here will generate a different list of available commands, including the very useful-looking <code>intron_shorter_than</code>. Typing this will take you deeper still, into the <code>GAG REMOVE intron_shorter_than&gt; </code> subconsole. Yes, that is a thing. From here, you can view the current minimum intron length ('0' means no minimum) or set a new minimum intron length. Once you perform the latter (or type <code>home</code>), you'll be returned to the main console.</p>

<h3><a name="i-hate-nested-subconsoles" class="anchor" href="#i-hate-nested-subconsoles"></a>...But I <em>Hate</em> Nested Subconsoles</h3>
<p>Relax, and be nice. We've got you covered. The second command described above ("flag any gene longer than 10,000 base pairs") could be carried out subconsole by subconsole, by typing <code>flag</code> and hitting enter to reach the <code>GAG FLAG&gt; </code> subconsole. From there you could type <code>gene_longer_than</code> and hit enter, and then type <code>10000</code> when prompted. You could. But who has the time?</p>
<p>If you know what you're doing, you can simply type 
<pre><code>GAG &gt; flag gene_longer_than 10000</code></pre>
from the main console. It does the same thing. Please note, though, that tab completion won't work past the first command, and spelling counts. If you make a mistake, GAG will take you as far as it can. For example, if you type 
<pre><code>GAG&gt; flag gene_longer_than ... IDK, what do you think?</code></pre> 
then GAG will take you as far as the <code>GAG FLAG gene_longer_than&gt; </code> subconsole, who will gently chide you for entering an invalid length.</p>
<p>Remember that should you get lost in some dark, scary subconsole, you can always type <code>home</code> to return to where you started.</p>

      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">GAG - Genome Annotation Generator maintained by <a href="https://github.com/BrianReallyMany">BrianReallyMany</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

  </body>
</html>
